### ISSUE TITLE
Support schema extension for GraphQL and REST data sources

### ISSUE THREAD

**Issue by alasdairtran** [wundergraph/wundergraph#284]
## Context

Consider the following. We're storing image metadata in our Postgres database using the PostGIS extension. Hasura supports PostGIS natively and creates a custom GraphQL scalar called `geography` to store the GPS coordinates. The value of this type is a GeoJSON. See https://hasura.io/docs/latest/schema/postgres/postgresql-types/#geography

We use Hasura as a backend for WunderGraph. Using the schema generated by WunderGraph, suppose I now try to query an image, e.g.
```
query GetImages {
  hasura_images {
    id
    location
  }
}
```
where `location`  has `geography` type. Then WunderGraph would complain that returning `location` field in the response has incorrect type:
```
ERROR ResolveGraphQLResponse for query failed {"component": "@wundergraph/node", "error": "invalid value type 'object' for path /data/hasura_images/1/location, expecting string, got: {\"type\":\"Point\",\"crs\":{\"type\":\"name\",\"properties\":{\"name\":\"urn:ogc:def:crs:EPSG::4326\"}},\"coordinates\":[-0.136667,50.818889]}. You can fix this by configuring this field as Int/Float Scalar"}
```
This kinda makes sense as Hasura returns a GeoJSON object for `location` but WunderGraph can't know that from introspection alone. The generated `wundergraph.api.schema.graphql` simply has:
```
scalar hasura_geography
```
So it seems we want to override the `geography` type generated by WunderGraph.

## Feature Request

It seems that this is what we want: https://docs.wundergraph.com/docs/guides/extend-json-fields-with-more-specific-types

But currently, it's only exposed to database sources.

It would be great if we can also extend GraphQL schemas from GraphQL sources (and maybe REST as well).

**Comment by alasdairtran** [comment]
Hi @YuriBuerov, I saw your PR #302 merged into master. Thanks a lot for coming up with a solution quickly. 

But it doesn't quite fully resolve this issue. Using the same example as above, Hasura expects `location` in the query to not have any subfield selection (because it is a custom scalar):

```graphql
query GetImages {
  images {
    id
    location
  }
}
```

and Hasura might return the following (note how Hasura automatically returns a resolved object for the `location` scalar, see [the official Hasura docs on GeoJSON](https://hasura.io/docs/latest/schema/postgres/postgresql-types/#geography)):

```json
{
  "data": {
    "images": [
      {
        "id": "713a2e37-702e-47d9-bc15-3ab699a0b2e0",
        "location": {
          "type": "Point",
          "crs": {
            "type": "name",
            "properties": {
              "name": "urn:ogc:def:crs:EPSG::4326"
            }
          },
          "coordinates": [
            -0.136667,
            50.818889
          ]
        }
      }
    ]
  }
}
```

If I extend the schema using the solution provided in #302, for e.g. like the following in `wundergraph.config.ts`:

```js
const hasura = introspect.graphql({
  id: 'hasura',
  apiNamespace: 'hasura',
  url: new EnvironmentVariable('HASURA_API_URL'),
  schemaExtension: `
    type CRSProperty {
      name: String!
    }
    type CRS {
      type: String!,
      properties: CRSProperty!
    }
    type Location {
      type: String!,
      crs: CRS!
      coordinates: [Float!]!
    }
  `,
  replaceCustomScalarTypeFields: [
    {
      entityName: 'images',
      fieldName: 'location',
      responseTypeReplacement: 'Location',
    },
  ],
});
```

And if I now use the original `GetImages`, then WunderGraph (or rather `graphql-go-tools`) would complain that 

```
ERROR: Error parsing operation GetImages: Field "location" of type "hasura_Location" must have a selection of subfields. Did you mean "location { ... }"?
```

If I instead add subfield selection for `location` to make WunderGraph happy:

```graphql
query GetImagesWithSubfieldSelection {
  images {
    id
    location {
      type
      crs {
        type
        properties {
          name
        }
      }
      coordinates
    }
  }
}
```

then Hasura would complain with the following:

```
{"errors":[{"message":"unexpected subselection set for non-object field","extensions":{"code":"validation-failed","path":"$.selectionSet.mochi_images.selectionSet.location"}},{"message":"unable to resolve","locations":[{"line":2,"column":3}]}],"data":null}
```

To summarize, I think the problem is this: **there are GraphQL sources that might return an already resolved complex object for a custom scalar type. And currently, there is no way to query these scalars in WunderGraph**. Maybe, the simplest solution might be to allow the user to disable validation for custom scalar fields. What do you think?

**Comment by jensneuse** [comment]
@YuriBuerov I think there's something wrong in the implementation. The original implementation that only worked for the Prisma datasource "marked" the custom scalars fields in a special way. When sending the query upstream, we need to dismiss the selections on the field, treating it like a scalar again. The Prisma datasource does this. I think we've missed to port this over.

@alasdairtran Once we've ported this over, it should work as expected. I'm just not sure why it happens, don't we have a test with Hasura?

**Comment by OLingard** [comment]
Hi @jensneuse, building on this comment you made:

> When sending the query upstream, we need to dismiss the selections on the field, treating it like a scalar again.

Can this also be applied to the `customIntScalars` implementation as well? Our upstream service has a custom scalar which is treated like an Int field so I tried to add the GraphQLIntrospection option of:

```ts
introspect.graphql({
  url: "xyz.com",
  customIntScalars: [
    'TestScalar'
  ]
})
```

This allowed me to treat this scalar as an Int field however in our upstream service we now get the error:
`Validation error of type VariableTypeMismatch: Variable type 'Int!' doesn't match expected type 'TestScalar!' @ 'TestType'`

Am I being naive to think that these two issues are related? It feels to me however that if we make any changes to the schema through Wundergraph we should dismiss these once we push them to the upstream source. 

For reference, when I remove the custom scalar block, the error I receive is:
```sh
2022-11-09T12:36:21Z ERROR apihandler/apihandler.go:773 ResolveGraphQLResponse {"component": "@wundergraph/node", "error": "invalid value type 'number' for path /data/TestType/TestField, expecting string, got: 12345. You can fix this by configuring this field as Int/Float Scalar"}
```

**Comment by jensneuse** [comment]
Ha! good catch. The problem here is that we treat this variable now as an Int, but before sending it to the Origin, we have to rewrite the Int to TestScalar. This is already implemented for another feature. What's missing is that we add "TestScalar" to the mapping table.

**Comment by OLingard** [comment]
Aah okay, that's great @jensneuse thanks. I've only started exploring Wundergraph today so I'm not in a position to make the changes just yet. Is there a process for reporting these bugs and asking for help drafting a pull request to solve the issue?

**Comment by jensneuse** [comment]
> Aah okay, that's great @jensneuse thanks. I've only started exploring Wundergraph today so I'm not in a position to make the changes just yet. Is there a process for reporting these bugs and asking for help drafting a pull request to solve the issue?

If you want to try creating a PR, here's what needs to be done:

from `replaceCustomScalars`, collect a list of `TypeFieldConfiguration`
https://github.com/wundergraph/wundergraph/blob/68b34d7b5436dfe8d88c7e3a27e0975a263e5034/packages/sdk/src/transformations/shema.ts#L4

Which needs to be added to the type configs from namespacing:

https://github.com/wundergraph/wundergraph/blob/68b34d7b5436dfe8d88c7e3a27e0975a263e5034/packages/sdk/src/definition/graphql-introspection.ts#L160

And looks something like this:

https://github.com/wundergraph/wundergraph/blob/68b34d7b5436dfe8d88c7e3a27e0975a263e5034/packages/sdk/src/definition/namespacing.test.ts#L367

So essentially, for all scalars you "rewrite", add them to the list to re-rename them, before sending to the origin. Then, when you're using a Variable Definition, it's rewritten automatically before sent upstream, see here:

https://github.com/wundergraph/graphql-go-tools/blob/df52582c3eaf2041b25ba31f15fcfb241e16ae10/pkg/engine/datasource/graphql_datasource/graphql_datasource.go#L147

So in the "planner" we lookup the mappings and "rewrite" the origin type name.
It sounds a bit complex, but is not that hard once you get your head around it.

Please repond if you'd like to work on this. Otherwise we'll pick it up some time in the future...

**Comment by OLingard** [comment]
Hi @jensneuse, is there a time box which this would need to be complete within? Happy to look at it so I understand the inner workings more however I've got holidays booked in for most of the week so I'd be unable to get round to it to mid next week. Aware that this breaks this functionality so if it is more urgent feel free to take it off my hands!

**Comment by OLingard** [comment]
Okay, I'll have a play about with it next week and see if I can get my head around the inner workings. I'll keep you updated and buzz with any questions. Thanks for the support!

**Comment by Slickstef11** [comment]
@OLingard wanted to check in and see if you had any questions? Have you been able to do a deep dive on the inner workings?

**Comment by OLingard** [comment]
Hi @Slickstef11, sorry had a busy week! I've added a ticket to our backlog which I'll pick up next week to look at this

**Comment by OLingard** [comment]
Hi @jensneuse, I've started the change in the Wundergraph codebase. You can find the current state here: https://github.com/wundergraph/wundergraph/compare/main...OLingard:wundergraph-test:FEAT/argument-replacements?expand=1.
I thought the current state would be sufficient to fix the bug with the `customIntScalar` however, there seems to be a hidden issue. Unfortunately with my inexperience with the codebase I know that finding this bug would take me an unnecessarily long amount of time to find and fix. This combined with a pile up of work I've got on my plate means it could be quicker for all if you please could take a look and see if you can find the issue? Happy to work with you on it or just look at the changes once you've made them.

Apologies for dropping this on you. Let me know your thoughts!

**Comment by jensneuse** [comment]
> Hi @jensneuse, I've started the change in the Wundergraph codebase. You can find the current state here: https://github.com/wundergraph/wundergraph/compare/main...OLingard:wundergraph-test:FEAT/argument-replacements?expand=1. I thought the current state would be sufficient to fix the bug with the `customIntScalar` however, there seems to be a hidden issue. Unfortunately with my inexperience with the codebase I know that finding this bug would take me an unnecessarily long amount of time to find and fix. This combined with a pile up of work I've got on my plate means it could be quicker for all if you please could take a look and see if you can find the issue? Happy to work with you on it or just look at the changes once you've made them.
> 
> Apologies for dropping this on you. Let me know your thoughts!

The link doesn't seem to work. Can you update it?

**Comment by OLingard** [comment]
> > Hi @jensneuse, I've started the change in the Wundergraph codebase. You can find the current state here: https://github.com/wundergraph/wundergraph/compare/main...OLingard:wundergraph-test:FEAT/argument-replacements?expand=1. I thought the current state would be sufficient to fix the bug with the `customIntScalar` however, there seems to be a hidden issue. Unfortunately with my inexperience with the codebase I know that finding this bug would take me an unnecessarily long amount of time to find and fix. This combined with a pile up of work I've got on my plate means it could be quicker for all if you please could take a look and see if you can find the issue? Happy to work with you on it or just look at the changes once you've made them.
> > Apologies for dropping this on you. Let me know your thoughts!
> 
> The link doesn't seem to work. Can you update it?

The PR above contains the changes!

**Comment by alasdairtran** [comment]
Is it possible to keep this issue open? I think PR #427 is still open.
